Imports.md

```diff
@@ -0,0 +1,407 @@
+
+An item import imports a single item. The item can be renamed with the `as` keyword.
+
+An import collection imports multiple items, and allows for nested imports.
+
+## Examples
+
+To compare it to the more widely known Typescript syntax, here are some examples.
+
+<table>
+<thead>
+<tr>
+<th>WGSL</th>
+<th>Typescript</th>
+</tr>
+</thead>
+<tbody>
+<tr>
+<td>
+
+```
+import ../geom/sphere/{draw, default_radius as foobar};
+```
+
+</td>
+<td>
+
+```ts
+import { draw, default_radius as foobar } from '../geom/sphere.wgsl';
+```
+
+</td>
+</tr>
+<tr>
+<td>
+
+```
+import bevy_ui/*;
+```
+
+</td>
+<td>
+
+```ts
+import * as bevy_ui from 'bevy_ui.wgsl';
+```
+
+</td>
+</tr>
+<tr>
+<td>
+
+```
+import bevy_pbr/{ 
+    forward_io/VertexOutput, 
+    pbr_types/{PbrInput, pbr_input_new}, 
+    pbr_bindings/* as pbr_b
+};
+```
+
+</td>
+<td>
+
+```ts
+import { VertexOutput } from 'bevy_pbr/forward_io.wgsl';
+import { PbrInput, pbr_input_new } from 'bevy_pbr/pbr_types.wgsl';
+import * as pbr_b from 'bevy_pbr/pbr_bindings.wgsl';
+```
+
+</td>
+</tr>
+</tbody>
+</table>
+
+## Parsing module.importable_item in the source code
+
+For tools that are parsing WGSL,
+here's how to extend the
+[WGSL grammar](https://www.w3.org/TR/WGSL/#grammar-recursive-descent)
+to parse imports:
+
+About extending the grammar:
+This one gets an additional meaning. A module property can also be a component.
+component_or_swizzle_specifier:
+| '.' member_ident component_or_swizzle_specifier ?
+
+This needs to be extended to support ident.ident
+for_init:
+| ident.ident func_call_statement.post.ident
+
+for_update:
+| ident.ident func_call_statement.post.ident
+
+global_decl:
+| attribute _ 'fn' ident ... '->' attribute _ ident.ident
+| 'alias' ident '=' ident.ident template_elaborated_ident.post.ident ';'
+
+primary_expression:
+| ident.ident template_elaborated_ident.post.ident
+| ident.ident template_elaborated_ident.post.ident argument_expression_list
+
+statement:
+| ident.ident template_elaborated_ident.post.ident argument_expression_list ';'
+
+type_specifier:
+| ident.ident ...
+
+## Behaviour-changing items
+
+These items cannot be imported, but they affect module behavior:
+
+- [Extensions](https://www.w3.org/TR/WGSL/#extensions)
+- [Global diagnostic filters](https://www.w3.org/TR/WGSL/#global-diagnostic-directive)
+
+These are always set at the very top in the main module, and affect all imported modules. They come before the imports.
+
+When, during parsing of imported modules, we encounter an extension or a global diagnostic filter, we check if the main module enables it, or sets the filter.
+If yes, everything is fine. If not, we throw an error.
+
+## Preserved items
+
+These items are preserved when importing a module. They are not imported, but will land in the final module with a mangled name.
```

Comment:
In naga-oil entrypoints are lowered to normal functions. I'm not sure if that's the correct semantics and is against the spec, but just wanted to comment on current behaviour that is being used in the wild.

Reply:
will add a note